### ISSUE TITLE
Major slowdown post 0.42.3

### ISSUE THREAD

**Issue by pd-l2ork** [nwjs/nw.js#7979]
### Issue Type

Before opening an issue, please search and see if it has already been raised.

- [x] Bug Report
- [ ] Feature Request

- [x] Successfully reproduced against the latest version of NW.js?

### Current/Missing Behavior

Most versions (tried everything between 0.42.4 up to 0.69.1, sometimes skipping a few versions) are significantly (10 or more times) slower than 0.42.3 and lower versions when drawing a busy SVG-based canvas. I am suspecting this may be due to nw2 features being enabled by default. If so, what is the highest version of nw.js that still supports --disable-features=nw2 switch?

If this is not due to nw2 flag enabled, what else could be the culprit? Is there some additional security layer running that I could disable since my app runs locally and connects via localhost proxy to the backend?

### Expected/Proposed Behavior

Speed of populating canvas should not be significantly different.

### Additional Info

- Operating System: any (tested on Linux, OSX, and Windows)
- NW.js Version: 0.42.4 and newer have problems
- Repro Link:
Unfortunately, in this case there is no simple way to make reproducible problem when it requires the full logic of a software that can create such a complex page. Best thing is to download my software (Pd-L2Ork) ideally on Windows or Linux where swapping nw.js versions is a matter of replacing a single folder inside installed application's folder. The app already ships with nw.js 0.42.3.

You can download it from: http://l2ork.music.vt.edu/main/make-your-own-l2ork/software/ (click on the "Complete (a.k.a. ‘Burrito Supreme’) Installer" section to expand it and download a version for the appropriate platform)

Replacing nw version: On Linux nw version is located in /usr/lib/pd-l2ork/bin/nw folder. On Windows in C:\Program Files (x86)\Pd-L2Ork\bin\nw. Mac is difficult since the files are all mixed, so I discourage testing on that one, even though results are the same. Just replace it with SDK (*not* runtime) of any other version by renaming that version's folder to nw and replacing this one. Then restart the app.

To open something graphically intense: start Pd-L2Ork (pd-l2ork on command line or double-click on the shortcut). Help (rightmost item in the menubar)->Help Browser->scroll to the bottom and click on L2Ork Tweeter and then again on L2Ork Tweeter (right below the CHANGELOG.txt).

Observe the time it takes to open and populate the window.
- Code snippet: N/A (I'll see if I can simply capture the canvas html code and make it into a page that loads at startup)
- Crash report:No crash, just a major slowdown.

**Comment by pd-l2ork** [comment]
Linked are index.html and package.json that is essentially the snapshot of how the screen looks when it loads.

https://l2ork.music.vt.edu/data/pd-l2ork/test.tar.gz

You can start the same using 0.42.3 and compare it to newer versions.

**Comment by pd-l2ork** [comment]
The difference is less obvious since there is no background communication (there is a lot of network code going back and forth to draw all this, but it is all identical in both cases). Still you can see the difference. I also spotted a lot of ssl errors on 0.42.3 being printed out in the console. Could it be that the 0.42.4 and onward use ssl layer for network proxy? If so, could this be the bottleneck?

**Comment by ayushmxn** [comment]
I'm seeing a black screen for v0.42.3 and v0.42.4 on Ubuntu 20. Not sure how to proceed

**Comment by pd-l2ork** [comment]
> I'm seeing a black screen for v0.42.3 and v0.42.4 on Ubuntu 20. Not sure how to proceed

@ayushmxn if you are running 0.42.4 or 0.42.5 (possibly also 0.42.6) on Linux. It indeed is broken and does not show anything. 0.42.3, however, should work just fine. You can also try 0.69.1.

**Comment by ayushmxn** [comment]
Could you send a screenshot of what a working example looks like? I'll try across different versions.

**Comment by ayushmxn** [comment]
v0.42.3 does not work for me. Could be a missing shared library which would explain why it work for you. 
```shell
[123138:123138:1015/205713.710040:ERROR:sandbox_linux.cc(371)] InitializeSandbox() called with multiple threads in process gpu-process.
[123138:123138:1015/205713.969746:ERROR:buffer_manager.cc(488)] [.DisplayCompositor]GL ERROR :GL_INVALID_OPERATION : glBufferData: <- error from previous GL command
```

**Comment by pd-l2ork** [comment]
Can you try 0.42.2 or earlier? Just about any lower works fine.

**Comment by pd-l2ork** [comment]
Here's how the window should look. This is on Win10 with 0.42.3. Running <path-to-nw-folder>nw . (from the folder where the two files provided in the aforesaid tarball are located)

![tweeter](https://user-images.githubusercontent.com/1066197/196016829-1497414f-976d-4d68-b67f-6ffa07b47531.png)

**Comment by ayushmxn** [comment]
I was able to reproduce!

**Comment by ayushmxn** [comment]
Putting down the diff for my reference:
nw diff: https://github.com/nwjs/nw.js/compare/b8c5658...5c1c6cf

**Comment by pd-l2ork** [comment]
@ayushmxn great! Please bear in mind that since 0.42.4 and 0.42.5 are effectively broken (do not display anything on my side), the slowdown may have only come into 0.42.6 (since there is no way to confirm that the problem is with 0.42.4).

**Comment by pd-l2ork** [comment]
> I was able to reproduce! There were quite a few changes made in between 0.42.3 and 0.42.4. I'll try to look into it.. Based on what I've seen, I may have a temp workaround. You could have 2 windows - one which shows a temporary [splash screen](https://github.com/nwutils/nw-splash-screen-example) / [circular progress](https://codepen.io/ayushmxn/pen/MWOLVGX) on start up until the main app is loaded.
> 
> Putting down the diff for my reference: nw diff: [b8c5658...5c1c6cf](https://github.com/nwjs/nw.js/compare/b8c5658...5c1c6cf) v8 diff: [nwjs/v8@00f437a...76eb2bb](https://github.com/nwjs/v8/compare/00f437a...76eb2bb)

How would a start-up screen solve the slowness? Or, would this be simply a distraction from the slowness? If so, that will not cut it, because this is a visual programming language and users can create things of varying complexity. While the aforesaid example may be seen as a more complex example, even more complex ones are certainly possible.

**Comment by ayushmxn** [comment]
> How would a start-up screen solve the slowness? Or, would this be simply a distraction from the slowness? If so, that will not cut it, because this is a visual programming language and users can create things of varying complexity. While the aforesaid example may be seen as a more complex example, even more complex ones are certainly possible.

AHh gotcha. I realise I misunderstood the problem when I replied above. Just reread the issue description:
> Most versions (tried everything between 0.42.4 up to 0.69.1, sometimes skipping a few versions) are significantly (10 or more times) slower than 0.42.3 and lower versions when drawing a busy SVG-based canvas.

**Comment by ayushmxn** [comment]
Could you send over a complex example of a *.pd file so I can observe the slowdown? I currently have a NW build running on my local machine which will probably take the whole of today to complete. I should be able to debug this during the coming weekend.

**Comment by pd-l2ork** [comment]
If you install pd-l2ork as per my first post (it has a deb installer) and follow instructions provided there (also copied below), it will open the very thing that I reduced to that html file above. And there the difference is dramatic:

You can download it from: http://l2ork.music.vt.edu/main/make-your-own-l2ork/software/ (click on the "Complete (a.k.a. ‘Burrito Supreme’) Installer" section to expand it and download a version for the appropriate platform).

Replacing nw version: On Linux nw version is located in /usr/lib/pd-l2ork/bin/nw folder. On Windows in C:\Program Files (x86)\Pd-L2Ork\bin\nw. Mac is difficult since the files are all mixed, so I discourage testing on that one, even though results are the same. Just replace it with SDK (not runtime) of any other version by renaming that version's folder to nw and replacing this one. Then restart the app.

To open something graphically intense: start Pd-L2Ork (pd-l2ork on command line or double-click on the shortcut). Go to Help (rightmost item in the menubar)->Help Browser->scroll to the bottom and click on L2Ork Tweeter and then again on L2Ork Tweeter (right below the CHANGELOG.txt).

**Comment by pd-l2ork** [comment]
@tharatau I was able to change the way my app works to be able to record the profile inside dev tools before the patch opens. Below are 2 profile files that I was able to save, one for 0.42.3 and one for 0.70.1:

https://l2ork.music.vt.edu/data/pd-l2ork/Profile-20221106T222201-0.42.3.json
https://l2ork.music.vt.edu/data/pd-l2ork/Profile-20221106T221257-0.70.1.json

Both run exactly the same code on my end. The only difference are the two nw.js versions.

I would really appreciate your assistance at dissecting what is causing these major slowdowns as I have zero experience with the profile option and this is currently a major issue preventing me from fixing crashes on OSX on 0.42.3 and earlier by upgrading to 0.60.1 or newer that fix this issue because anything post 0.42.3 that does run without crashing is ~10 times slower than older versions. Thank you.

**Comment by pd-l2ork** [comment]
Another update, having looked at the profile tab in the dev tools, it appears the major difference stems from the "scripting" aspect. This is the one that makes this huge difference in load times between the two. Whereas 0.42.3 takes barely a fraction of a second to process all the scripting, on 0.70.1 the same takes up 9 seconds (!).

**Comment by pd-l2ork** [comment]
@tharatau Found it! While before doing this would generate an error, now it finally works. If I disable nw2 by passing the argument --disable-features=nw2 the slowdown is gone. So, this is clearly a part of the nw2 introduction. Now onto figuring out what is causing such colossal slowdown (and this is also where I need to step back because I have no knowledge of the nw1 vs nw2)...

**Comment by tharatau** [comment]
@pd-l2ork Apologies for the late response. That's awesome. I'm not too sure how nw1 differs from nw2 also.

cc #7230 @rogerwang

**Comment by pd-l2ork** [comment]
@tharatau @rogerwang I strongly suggest that you do not obsolete nw1 until this major slowdown is resolved because it is significant (10x slower in nw2). Thank you for your consideration.

**Comment by pd-l2ork** [comment]
@rogerwang to simplify troubleshooting, you can also simply use latest nw.js and change the flag in the package.json that enables or disables nw2 to observe the dramatic change in performance. Please let me know what OS you run and I will make you a build of pd-l2ork for your platform to easily troubleshoot this.

Folder locations for different OS installs:
Linux: /usr/lib/pd-l2ork/bin/package.json
OSX: /Applications/Pd-L2Ork.app/Contents/Resources/app.nw/package.json
Windows: C:/Program Files (x86)/Pd-L2Ork/bin/package.json

**Comment by rogerwang** [comment]
I tried [your sample](https://github.com/nwjs/nw.js/issues/7979#issuecomment-1276961427) under nw1 and nw2 with the latest version, but I don't notice and difference...

**Comment by pd-l2ork** [comment]
> I tried [your sample](https://github.com/nwjs/nw.js/issues/7979#issuecomment-1276961427) under nw1 and nw2 with the latest version, but I don't notice any difference...

@rogerwang that is because it is a scripting issue as noted in my comment above. You should try the pd-l2ork app instead. See the two profile files I uploaded for comparison.

**Comment by pd-l2ork** [comment]
@rogerwang what OS are you using? I can prepare a version of the app that is easy to test and show the slowdown, including the ability to profile...

**Comment by rogerwang** [comment]
Please provide a minimal case to reproduce, not a whole application...

**Comment by pd-l2ork** [comment]
@rogerwang this is a complex application that does pretty much everything nwjs can, from networking, to scripting, drawing, and interacting. It clearly exhibits a huge slowdown when everything remains the same except for toggling the nw2 feature. How would I go about narrowing down the source of the problem? Is there a way to use profiling tool to be able to do this?

**Comment by rogerwang** [comment]
You can try using the devtools provided by Chromium, and insert code in various steps in your application to monitor the performance e.g. printing timestamps.

**Comment by pd-l2ork** [comment]
@rogerwang so, I added bunch of timestamps in hope of identifying one major slowdown. Instead, I got everything executing slower inside js. When there are a few calls, this is not noticeable, but as the number of calls increases, the overall execution time deviates. As an example, below are two files executing same select functions with the only difference being one disabling nw2 in the package.json file. I have not timestamped all functions (that would take quite an effort given the code is thousands of lines long), but this gives a nice cross-section. The text inserts =======\n at any point where there is a jump larger than 10ms. After executing little over 450 calls, nw2 is at 3848ms whereas nw1 is at 1484ms. All times shown are in ms. Both are using 0.67.1 due to #7981 bug in nw1 mode in newer versions. That said, the behavior is exactly the same on any version that has nw2 enabled by default (IIRC 0.42.4 onward).

Is there a chance that there is some debugging or logging stuff running somewhere in the background that is causing this to bleed slowdown across the board?

[nw2-disabled-log](https://l2ork.music.vt.edu/data/pd-l2ork/nw1)
[nw2-enabled-log](https://l2ork.music.vt.edu/data/pd-l2ork/nw2)

**Comment by agraef** [comment]
With #7230 looming large, I thought I'd chime in here with a few remarks, because the issue affects me, too. nw.js is an awesome project, but if the JS performance issue can't be fixed, and NW1 mode goes away, then some projects will likely be stuck with nw.js 0.67.1 forever, as that's the last version not suffering from #7981. That's a real bummer. :(

I'm having the same issue as @pd-l2ork reported with Jonathan Wilke's [Purr Data](https://github.com/agraef/purr-data). (Which is pd-l2ork's sister project, so this isn't surprising.) There's a lot of JavaScript execution going on in that app, you really can't demonstrate the issue with a minimal example. I have a little patch (basically a Pd object written in Lua which draws a graphic animation inside an svg element in an nw.js window) which demonstrates the issue.

Here's how the patch looks like in action (the gif  looks a bit sluggish, but with NW2 disabled the animation really looks a lot smoother in real life):

![osci3d-example](https://github.com/user-attachments/assets/3ca9bb2f-a3c8-4308-a205-c65cba16a536)

The Lua script does a lot of calls into JavaScript via C code in the Pd engine to modify the svg element in the patch window in real-time (basically using eval to evaluate simple Java function calls). So the JS interpreter has its work cut out. It still works fine with NW2 disabled, but with NW2 the entire application becomes sluggisch because it spends so much time in the JS interpreter.

Here's what the dev tool performance monitor gives with NW2 disabled:
![nw1-js-performance](https://github.com/user-attachments/assets/28bd5aaa-225b-499c-b41b-661cc91a37ad)

And here's the graph with NW2 enabled:
![nw2-js-performance](https://github.com/user-attachments/assets/7b1d4d39-dac5-4684-9c66-4896ff59fa2b)

As you can see, the time spent in JS execution went up by almost an order of magnitude, like @pd-l2ork observed. The graphics rendering itself is fine (I originally thought that was the issue). Maybe eval being used for dynamic JS code execution is the issue here? Whatever the reason is, NW is *much* slower in this use case.

Now I have to admit that this is very far from your run-of-the-mill use case for a browser engine. :) But with NW1 it certainly works extremely well, and a lot of work went into doing that GUI port (upstream Pd is still using Tcl/Tk for its GUI, which is much less advanced). So I *really* hope that the JS execution issues can be sorted out in NW2, and that NW1 mode is kept around until that time, so that we can continue using newer nw.js versions than what is now possible.

I'm afraid that it will be hard to replicate the issue in an example with just a basic nw.js app, but I can post download links for Purr Data and the patch example, if someone is still willing to look into this issue.

**Comment by agraef** [comment]
@rogerwang On second thought, it ought to be possible to demonstrate this kind of major slowdown in the JS performance from NW1 to NW2 with a simpler test app which dynamically modifies a document with some svg graphics, using a relatively high framerate, and does lots of dynamic JS code execution via eval to do it. That's basically the gist of what our examples here do.

AFAICT you don't do this kind of performance measurement in your existing test framework, but the effect should be readily noticable by just running the same app with NW1 and NW2, respectively.

I'll try to cobble something together as soon as I find the time. If there is such a major degradation in JS performance, you surely want to fix it before phasing out NW1, no? ;-) As you know the sources better than anyone else, maybe you can also think of a reason why the switch from NW1 to NW2 could have such an impact on the JS interpreter?

**Comment by rogerwang** [comment]
Unfortunately the NW1 mode was deprecated long time ago. It's unmaintained and could break at any time...

**Comment by agraef** [comment]
@rogerwang, No-one wants to drag along legacy code forever, I understand that. :)

But that means that we're stuck between a rock and a hard place, as long as the JS performance issue we observe remains unfixed. Are you willing to look into it, if I provide a smaller example demonstrating it?

**Comment by rogerwang** [comment]
> Are you willing to look into it, if I provide a smaller example demonstrating it?

I asked for that 2 years ago ... https://github.com/nwjs/nw.js/issues/7979#issuecomment-1309142267

**Comment by agraef** [comment]
Yes, I understand, but I didn't actually notice the issue (or this bug report) myself until very recently. Ok, I'll give it a go then. Thanks for your reply.

**Comment by GnorTech** [comment]
We tested SVG rendering performance with 2000 elements on the latest version — NW.js and Chrome show identical performance (13.5ms vs 13.6ms). A minimal reproduction was requested over 2 years ago but was never provided. Closing — please reopen with a standalone test case if the issue persists.
